### PR TITLE
Fix setting PYTHONPATH on Python 2 not working

### PR DESCRIPTION
**Pull request by jd**
- ensure we do not rewrite the ``PYTHONPATH``
- ensure we respect the ``-E`` flag.

### CODE REVIEW

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py [hunk 1]
```diff
@@ -114,6 +108,45 @@ def abs_path(value):
...
+        if not skip_rewrite:
+            sys.path[at] = translate_directory(value)
+
+
+def revert_pythonpath_rewrite():
+    # Now that we can access os.environ["PYTHONPATH"], we can revert our
+    # rewrite of thoses paths done in `rewrite_standard_library_sys_path`.
+    import os
+
+    python_paths = os.environ.get("PYTHONPATH", "").split(":")
```
Comment (gaborbernat) [inline src/virtualenv/create/via_global_ref/builtin/python2/site.py]: fairly certain these should be absolute paths too 👍

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py [hunk 2]
```diff
@@ -114,6 +108,45 @@ def abs_path(value):
...
+    for at, value in enumerate(sys.path):
+        value = abs_path(value)
+        # replace old sys prefix path starts with new
+        skip_rewrite = value == exe_dir  # don't fix the current executable location, notably on Windows this gets added
+        skip_rewrite = skip_rewrite  # ___SKIP_REWRITE____
+        if not skip_rewrite:
+            sys.path[at] = translate_directory(value)
+
+
+def revert_pythonpath_rewrite():
```
Comment (gaborbernat) [inline src/virtualenv/create/via_global_ref/builtin/python2/site.py]: I prefer python path to be written as ``python_path``

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py [hunk 3]
```diff
@@ -114,6 +108,45 @@ def abs_path(value):
     return value
 
 
+def translate_directory(directory):
```
Comment (gaborbernat) [inline src/virtualenv/create/via_global_ref/builtin/python2/site.py]: this should be after ``rewrite_standard_library_sys_path`` to ordered top-down by the occurrence

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py [hunk 4]
```diff
@@ -6,15 +6,29 @@
 """
 import sys
 
+base_executable = None
```
Comment (gaborbernat) [inline src/virtualenv/create/via_global_ref/builtin/python2/site.py]: please don't, don't introduce globals they just make code hard to follow

Comment (jd) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: Ok, though the list of arguments is going to be long to carry everwhere :(

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py (hunk 2 above)
Comment (jd) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: I'm not sure it's going to make the function name clear that we're talking about `PYTHONPATH` though. You sure?

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py (hunk 1 above)
Comment (jd) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: Fixed

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py (hunk 4 above)
Comment (gaborbernat) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: I personally prefer that over globals. Also, we don't have to carry everything everywhere.

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py (hunk 2 above)
Comment (gaborbernat) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: yes

Comment (gaborbernat) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: yes

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py (hunk 1 above)
Comment (gaborbernat) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: I don't see the change, conversations are usually resolved by the one starting the discussion.

Comment (gaborbernat) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: I don't see. The conversation is resolved by the one starting the conversation for this reason.

**Comment by jd** [comment]
Is there anything left that needs fixing?

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py [hunk 5]
```diff
@@ -13,8 +13,14 @@ def main():
...
     sys.base_exec_prefix = config["base-exec-prefix"]
     sys.base_executable = config["base-executable"]
+    base_executable = abs_path(sys.base_executable)
+    base_prefix, base_exec_prefix = abs_path(sys.base_prefix), abs_path(sys.base_exec_prefix)
+    exe = abs_path(sys.executable)
+    exe_dir = exe[: exe.rfind(sep)]
+    prefix, exec_prefix = abs_path(sys.prefix), abs_path(sys.exec_prefix)
     global_site_package_enabled = config.get("include-system-site-packages", False) == "true"
-    rewrite_standard_library_sys_path()
+    rewrite_standard_library_sys_path(base_executable, exe_dir, exec_prefix, base_prefix, prefix, base_exec_prefix)
```
Comment (gaborbernat) [inline src/virtualenv/create/via_global_ref/builtin/python2/site.py]: Feels to me we only need all these variables here because we're doing the rewrite and revert in two functions, cant' we just do it in one to avoid this extra overhead?

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py [hunk 6]
```diff
@@ -60,6 +66,10 @@ def load_host_site():
 sep = "\\" if sys.platform == "win32" else "/"  # no os module here yet - poor mans version
 
 
+def dir_starts_with(directory, prefix):
```
Comment (gaborbernat) [inline src/virtualenv/create/via_global_ref/builtin/python2/site.py]: this is out of order, usually, functions are defined after they are called

Comment (gaborbernat) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: so they read top down

Comment (jd) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: You should invest in some automated checking tools. Bothering contributors with such trivial details on each review is kinda frustrating.

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py (hunk 1 above)
Comment (jd) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: See L140

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py (hunk 5 above)
Comment (jd) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: Sure

**Comment by jd** [comment]
Could you give me some hints about where to add some tests?
I took a look at https://github.com/pypa/virtualenv/pull/1641 but it's unclear to me how you added/changed the tests.

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py (hunk 6 above)
Comment (gaborbernat) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: That's why we have pre commit. I haven't got a tool for checking this yet though.

**Comment by gaborbernat** [comment]
In that pr the test were adding brew Python to the ci matrix. Here you want a test inside test_create 👍set Python path, run create, check sys path of the debug property 👍

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py (hunk 6 above)
Comment (gaborbernat) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: That being said feel free to take this as far as you're willing to. I'll pick up and clean things up once you're done, and ship it. As a maintainer, I feel like I need to guide the code to have some unified style and ensure PRs don't just fix issues but leave the codebase in a maintainable state. Hope you understand.

Comment (jd) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: You need a CI check, not (just) pre-commit.

If your plan is to manually guide people to your style and pick up behind people, you're going to be exhausted before long and even might lose contributors in the process. Which is not something I wish you!

Comment (gaborbernat) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: As I said I'm not aware of a tool that does this at the moment. Pre commit is run part of the ci. So a pre commit test is a ci test.

**Comment by jd** [comment]
I'm blocked by https://github.com/pypa/virtualenv/issues/1683 to try locally to add tests :(

**Comment by jd** [comment]
Any idea how you set UTF-8 strings in a Windows env? 🤔

**Comment by pfmoore** [comment]
Environment variables are Unicode in Windows, so I'm not quite sure the question makes sense. Having said that, if this is Python 2, `os.environ` probably uses the ANSI APIs to read the environment, so the practical answer would likely be "you can't - you always get the default encoding".

**Comment by jd** [comment]
My question is related to the CI failure. IIUC the dir contains Unicode chars and setting that as `PYTHONPATH` is a problem… :(

**Comment by pfmoore** [comment]
Ah, OK. So passing a general Unicode value through to the subprocess' environment. I don't think you can, in Python 2 (IIRC, you can only use characters that are valid in the system ANSI codepage) 🙁 (I may be wrong, it's a long while since I used Python 2).

Sorry I can't be more help.

File: tests/conftest.py [hunk 7]
```diff
@@ -311,6 +311,28 @@ def change_env_var(key, value):
...
+def override_env_var():
+
+    original_env = {}
+
+    def _environ_override(var, value):
+        original_env.setdefault(var, os.environ.get(var, _ENV_EMPTY))
+        os.environ[var] = value
+
+    try:
+        yield _environ_override
```
Comment (gaborbernat) [inline tests/conftest.py]: any reason not just use the builtin ``monkeypatch.setenv``?

Comment (jd) [reply tests/conftest.py]: No, I just didn't know about it!

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py [hunk 8]
```diff
@@ -78,40 +78,59 @@ def read_pyvenv():
...
-                value = "{}{}".format(orig_exe_folder, value[len(exe_dir) :])
-            elif value.startswith(prefix):
-                value = "{}{}".format(base_prefix, value[len(prefix) :])
-            elif value.startswith(exec_prefix):
-                value = "{}{}".format(base_exec_prefix, value[len(exec_prefix) :])
-        sys.path[at] = value
+            sys.path[at] = map_path(path, base_exe, exe_dir, exec_prefix, base_prefix, prefix, base_exec)
+
+    # the rewrite above may have changed elements from PYTHONPATH, revert these if on
+    if sys.flags.ignore_environment:
```
Comment (gaborbernat) [inline src/virtualenv/create/via_global_ref/builtin/python2/site.py]: @jd added handling for the ``-E`` flag

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py [hunk 9]
```diff
@@ -78,40 +78,59 @@ def read_pyvenv():
...
+        return
+    import os
+
+    python_paths = []
+    if "PYTHONPATH" in os.environ and os.environ["PYTHONPATH"]:
+        for path in os.environ["PYTHONPATH"].split(os.pathsep):
+            if path not in python_paths:
+                python_paths.append(path)
+    sys.path[: len(python_paths)] = python_paths
```
Comment (gaborbernat) [inline src/virtualenv/create/via_global_ref/builtin/python2/site.py]: to supports setting paths both within the host prefix and the venv prefix instead of trying to fix up let's just overwrite the rewritten part that belongs to python path

File: tests/unit/create/test_creator.py [hunk 10]
```diff
@@ -386,3 +388,58 @@ def test_create_distutils_cfg(creator, tmp_path, monkeypatch):
 
     package_folder = result.creator.platlib / "demo"  # prefix is set to the virtualenv prefix for install
     assert package_folder.exists()
+
+
+@pytest.mark.parametrize("python_path_on", [True, False], ids=["on", "off"])
+@pytest.mark.skipif(PY3, reason="we rewrite sys.path only on PY2")
+def test_python_path(monkeypatch, tmp_path, python_path_on):
```
Comment (gaborbernat) [inline tests/unit/create/test_creator.py]: have a dedicated test for this use case

File: tests/unit/create/test_creator.py [hunk 11]
```diff
@@ -386,3 +388,58 @@ def test_create_distutils_cfg(creator, tmp_path, monkeypatch):
...
+        cmd.extend(["-c", "import json; import sys; print(json.dumps(sys.path))"])
+        print(cmd)
+        return json.loads(subprocess.check_output(cmd))
+
+    monkeypatch.delenv(str("PYTHONPATH"), raising=False)
+    base = _get_sys_path()
+
+    # note the value result.creator.interpreter.system_stdlib cannot be set, as that would disable our custom site.py
+    python_paths = [
+        str(Path(result.creator.interpreter.prefix)),
```
Comment (gaborbernat) [inline tests/unit/create/test_creator.py]: test various edge cases for the python path as just one path within the venv

File: tests/unit/create/test_creator.py [hunk 12]
```diff
@@ -386,3 +388,58 @@ def test_create_distutils_cfg(creator, tmp_path, monkeypatch):
...
+        assert not (set(base) - set(extra_all))  # all base paths are present
+        abs_python_paths = list(OrderedDict((os.path.abspath(ensure_text(i)), None) for i in python_paths).keys())
+
+        extra_as_python_path = extra_all[: len(abs_python_paths)]
+        assert abs_python_paths == extra_as_python_path  # python paths are there at the start
+
+        non_python_path = extra_all[len(abs_python_paths) :]
+        assert non_python_path == [i for i in base if i not in extra_as_python_path]
+    else:
+        assert base == extra_all
```
Comment (gaborbernat) [inline tests/unit/create/test_creator.py]: also, test that we respect ``-E``

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py (hunk 9 above)
Comment (jd) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: I'm a bit confused here. How can you be sure the first paths in `sys.path` can be replaced like that?

File: src/virtualenv/create/via_global_ref/builtin/python2/site.py [hunk 13]
```diff
@@ -78,40 +78,59 @@ def read_pyvenv():
...
+    # the rewrite above may have changed elements from PYTHONPATH, revert these if on
+    if sys.flags.ignore_environment:
+        return
+    import os
+
+    python_paths = []
+    if "PYTHONPATH" in os.environ and os.environ["PYTHONPATH"]:
+        for path in os.environ["PYTHONPATH"].split(os.pathsep):
+            if path not in python_paths:
+                python_paths.append(path)
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply src/virtualenv/create/via_global_ref/builtin/python2/site.py]: Keeping the order is very important, a set would loose that information

**Comment by gaborbernat** [comment]
@jd it's just how pythonpath is parsed at the interpreter level.

**Comment by gaborbernat** [comment]
Hello, a fix for this issue has been released via virtualenv 20.0.8; see https://pypi.org/project/virtualenv/20.0.8/ (https://virtualenv.pypa.io/en/latest/changelog.html#v20-0-8-2020-03-04). Please give a try and report back if your issue has not been addressed; if not, please comment here, and we'll reopen the ticket. We want to apologize for the inconvenience this has caused you and say thanks for having patience while we resolve the unexpected bugs with this new major release.
thanks
